File: webui/dashboard/view/viewDashboard.js
// This file is part of the Resultra project.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.


function loadDashboardView(pageLayout,databaseID, dashboardID) {


	hideSiblingsShowOne('#dashboardViewSidebarProps')
	hideSiblingsShowOne('#dashboardCanvas')
	
	// Initially hide the RHS sidebar. The sidebar is only shown when individual components are selected.
	theMainWindowLayout.hideRHSSidebar()
	

	var viewDashboardCanvasSelector = '#dashboardCanvas'
		
	viewDashboardContext = { 
				dashboardID:dashboardID,
		 		databaseID: databaseID} 			
	
	var getDashboardParams = { dashboardID: dashboardID }	

	jsonAPIRequest("dashboard/getProperties",getDashboardParams,function(dashboardInfo) {
		pageLayout.setCenterContentHeader(dashboardInfo.name)		
	})
				
				
	function initDashboardComponentViewBehavior($component,componentID, viewDashboardConfig) {

		var $parentDashboardCanvas = $(viewDashboardCanvasSelector)	
		initObjectSelectionBehavior($component, 
				$parentDashboardCanvas,function(selectedComponentID) {
			console.log("dashboard view object selected: " + selectedComponentID)
			var selectedObjRef	= getContainerObjectRef($component)
			viewDashboardConfig.selectionFunc($component,selectedObjRef)
			theMainWindowLayout.openRHSSidebar()
		})
	}		

	var loadDashboardConfig = {
		dashboardContext: viewDashboardContext,
		doneLoadingDashboardDataFunc: function() {

			initObjectCanvasSelectionBehavior(viewDashboardCanvasSelector, function() {
				// Hide the RHS sidebar when the main canvas of the dashboard is selected.
				theMainWindowLayout.hideRHSSidebar()
			})


		},
		initBarChartComponent: function($barChart,barChartRef) {

			var barChartViewConfig = barChartViewDashboardConfig(barChartRef)

			console.log("Init bar chart component")

			initDashboardComponentViewBehavior($barChart,
					barChartRef.barChartID,barChartViewConfig)
		},
		initSummaryTableComponent: function($summaryTable,summaryTableRef) {

			var summaryTableViewConfig = summaryTableViewDashboardConfig(summaryTableRef)

			console.log("Init summary table component")
			
			initDashboardComponentViewBehavior($summaryTable,
					summaryTableRef.summaryTableID,summaryTableViewConfig)

		},
		initHeaderComponent: function($header,headerRef) {

			var viewConfig = headerViewDashboardConfig(headerRef)

			console.log("Init header component")
			
		},
		initGaugeComponent: function($gauge,gaugeRef) {

			var viewConfig = gaugeViewDashboardConfig(gaugeRef)

			console.log("Init gauge component")
			
			initDashboardComponentViewBehavior($gauge,gaugeRef.gaugeID,viewConfig)

		},
		initSummaryValComponent: function($summaryVal,summaryValRef) {

			var viewConfig = summaryValViewDashboardConfig(summaryValRef)

			console.log("Init gauge component")
			
			initDashboardComponentViewBehavior($summaryVal,summaryValRef.summaryValID,viewConfig)

		}

	}

	loadDashboardData(loadDashboardConfig)
}

File: webui/dashboard/view/viewDashboard.go
// This file is part of the Resultra project.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
package view

import (
	"fmt"
	"net/http"
	"github.com/resultra/resultra/server/common/databaseWrapper"
	"github.com/resultra/resultra/server/databaseController"
	"github.com/resultra/resultra/server/userRole"
	"github.com/resultra/resultra/webui/dashboard/components"
)

type ViewDashboardInfo struct {
	DatabaseID      string
	DatabaseName    string
	DashboardID     string
	DashboardName   string
	CurrUserIsAdmin bool
	Title           string
	ComponentParams components.ComponentViewTemplateParams
}

func getViewDashboardInfo(r *http.Request, dashboardID string) (*ViewDashboardInfo, error) {

	trackerDBHandle, dbErr := databaseWrapper.GetTrackerDatabaseHandle(r)
	if dbErr != nil {
		return nil, dbErr
	}

	dashboardDbInfo, getErr := databaseController.GetDashboardDatabaseInfo(trackerDBHandle, dashboardID)
	if getErr != nil {
		return nil, getErr
	}

	hasViewPrivs, privsErr := userRole.CurrentUserHasDashboardViewPrivs(trackerDBHandle, r, dashboardDbInfo.DatabaseID, dashboardID)
	if privsErr != nil {
		return nil, privsErr
	}
	if !hasViewPrivs {
		return nil, fmt.Errorf("ERROR: No permissions to view this dashboard")
	}

	isAdmin := userRole.CurrUserIsDatabaseAdmin(r, dashboardDbInfo.DatabaseID)

	dashboardInfo := ViewDashboardInfo{
		DatabaseID:      dashboardDbInfo.DatabaseID,
		DatabaseName:    dashboardDbInfo.DatabaseName,
		DashboardID:     dashboardDbInfo.DashboardID,
		DashboardName:   dashboardDbInfo.DashboardName,
		CurrUserIsAdmin: isAdmin}

	return &dashboardInfo, nil

}
